File: r_code/R/multilayer_utils.R
#' @keywords internal
prepare_multilayer_graphs <- function(layers, directed = FALSE) {
  if (!requireNamespace("igraph", quietly = TRUE)) {
    stop("Package 'igraph' is required.", call. = FALSE)
  }

  if (!is.list(layers) || length(layers) < 2) {
    stop("`layers` must be a list with at least two network layers.", call. = FALSE)
  }

  graph_layers <- lapply(seq_along(layers), function(i) {
    layer <- layers[[i]]
    if (inherits(layer, "igraph")) {
      return(layer)
    }

    if (!is.matrix(layer)) {
      stop(sprintf("Layer %d is not an igraph object or adjacency matrix.", i), call. = FALSE)
    }

    if (nrow(layer) != ncol(layer)) {
      stop(sprintf("Layer %d adjacency matrix must be square.", i), call. = FALSE)
    }

    igraph::graph_from_adjacency_matrix(
      adjmatrix = layer,
      mode = if (directed) "directed" else "undirected",
      weighted = TRUE,
      diag = FALSE
    )
  })

  if (is.null(names(graph_layers))) {
    names(graph_layers) <- paste0("layer_", seq_along(graph_layers))
  }

  graph_layers
}

#' @keywords internal
make_layer_links <- function(n_layers, layer_links = NULL) {
  if (is.null(layer_links)) {
    return(data.frame(
      from = seq_len(n_layers - 1),
      to = seq(2, n_layers),
      weight = 1,
      stringsAsFactors = FALSE
    ))
  }

  if (!is.data.frame(layer_links) || !all(c("from", "to") %in% names(layer_links))) {
    stop("`layer_links` must be a data.frame with columns `from` and `to`.", call. = FALSE)
  }

  if (!"weight" %in% names(layer_links)) {
    layer_links$weight <- 1
  }

  if (any(layer_links$from < 1 | layer_links$to < 1 |
          layer_links$from > n_layers | layer_links$to > n_layers)) {
    stop("`layer_links` indices must be between 1 and number of layers.", call. = FALSE)
  }

  layer_links
}

#' @keywords internal
fit_layer_communities <- function(graph_layers, algorithm = c("louvain", "leiden"),
                                  resolution_parameter = 1,
                                  directed = FALSE) {
  algorithm <- match.arg(algorithm)

  lapply(graph_layers, function(g) {
    g_input <- g

    if (algorithm == "louvain") {
      if (directed && igraph::is_directed(g_input)) {
        g <- igraph::as.undirected(g_input, mode = "collapse", edge.attr.comb = list(weight = "sum", "ignore"))
      }
      cl <- igraph::cluster_louvain(g, weights = igraph::E(g)$weight)
    } else {
      cl <- igraph::cluster_leiden(
        g,
        objective_function = if (directed) "CPM" else "modularity",
        resolution_parameter = resolution_parameter,
        weights = igraph::E(g)$weight
      )
    }

    list(
      membership = igraph::membership(cl),
      modularity = if (igraph::is_directed(g)) NA_real_ else igraph::modularity(g, igraph::membership(cl),
                                      weights = igraph::E(g)$weight),
      communities = split(seq_along(igraph::membership(cl)), igraph::membership(cl))
    )
  })
}

#' @keywords internal
weighted_jaccard <- function(a, b) {
  inter <- length(intersect(a, b))
  union <- length(union(a, b))
  if (union == 0) {
    return(0)
  }
  inter / union
}

#' @keywords internal
weighted_overlap <- function(a, b) {
  inter <- length(intersect(a, b))
  min_size <- min(length(a), length(b))
  if (min_size == 0) {
    return(0)
  }
  inter / min_size
}

#' @keywords internal
community_overlap_edges <- function(fit, layer_links, metric = c("jaccard", "overlap"),
                                    min_similarity = 0) {
  metric <- match.arg(metric)
  sim_fun <- if (metric == "jaccard") weighted_jaccard else weighted_overlap

  edge_rows <- list()

  for (i in seq_len(nrow(layer_links))) {
    from_idx <- layer_links$from[i]
    to_idx <- layer_links$to[i]
    layer_weight <- layer_links$weight[i]

    from_comms <- fit[[from_idx]]$communities
    to_comms <- fit[[to_idx]]$communities

    for (from_c in seq_along(from_comms)) {
      for (to_c in seq_along(to_comms)) {
        sim <- sim_fun(from_comms[[from_c]], to_comms[[to_c]])
        weighted_sim <- sim * layer_weight

        if (weighted_sim >= min_similarity) {
          edge_rows[[length(edge_rows) + 1]] <- data.frame(
            from_layer = from_idx,
            to_layer = to_idx,
            from_community = from_c,
            to_community = to_c,
            similarity = sim,
            layer_weight = layer_weight,
            weighted_similarity = weighted_sim,
            stringsAsFactors = FALSE
          )
        }
      }
    }
  }

  if (length(edge_rows) == 0) {
    return(data.frame(
      from_layer = integer(0),
      to_layer = integer(0),
      from_community = integer(0),
      to_community = integer(0),
      similarity = numeric(0),
      layer_weight = numeric(0),
      weighted_similarity = numeric(0)
    ))
  }

  do.call(rbind, edge_rows)
}


#' @keywords internal
add_community_self_loops <- function(edge_df, fit, layer_links,
                                     self_loop_multiplier = 1,
                                     min_similarity = 0) {
  loop_rows <- list()

  for (i in seq_len(nrow(layer_links))) {
    layer_idx <- layer_links$from[i]
    layer_weight <- layer_links$weight[i]
    comms <- fit[[layer_idx]]$communities

    for (comm_idx in seq_along(comms)) {
      weighted_sim <- 1 * layer_weight * self_loop_multiplier
      if (weighted_sim >= min_similarity) {
        loop_rows[[length(loop_rows) + 1]] <- data.frame(
          from_layer = layer_idx,
          to_layer = layer_idx,
          from_community = comm_idx,
          to_community = comm_idx,
          similarity = 1,
          layer_weight = layer_weight,
          weighted_similarity = weighted_sim,
          stringsAsFactors = FALSE
        )
      }
    }
  }

  if (length(loop_rows) == 0) {
    return(edge_df)
  }

  rbind(edge_df, do.call(rbind, loop_rows))
}

File: python_code/src/dynamic_multiplex/multilayer_utils.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable

import networkx as nx
import numpy as np
import pandas as pd


@dataclass
class LayerCommunityFit:
    membership: dict[int, int]
    modularity: float | None
    communities: dict[int, list[int]]


def _as_graph(layer, directed: bool) -> nx.Graph | nx.DiGraph:
    if isinstance(layer, (nx.Graph, nx.DiGraph)):
        return layer

    matrix = np.asarray(layer)
    if matrix.ndim != 2 or matrix.shape[0] != matrix.shape[1]:
        raise ValueError("Each layer must be a square adjacency matrix or a NetworkX graph.")

    graph_type = nx.DiGraph if directed else nx.Graph
    graph = nx.from_numpy_array(matrix, create_using=graph_type)
    return graph


def prepare_multilayer_graphs(layers: list, directed: bool = False) -> list[nx.Graph | nx.DiGraph]:
    if not isinstance(layers, list) or len(layers) < 2:
        raise ValueError("`layers` must be a list with at least two network layers.")

    return [_as_graph(layer, directed=directed) for layer in layers]


def make_layer_links(n_layers: int, layer_links: list[dict] | pd.DataFrame | None = None) -> pd.DataFrame:
    if layer_links is None:
        return pd.DataFrame(
            {
                "from": np.arange(1, n_layers),
                "to": np.arange(2, n_layers + 1),
                "weight": 1.0,
            }
        )

    links = pd.DataFrame(layer_links).copy()
    if not {"from", "to"}.issubset(links.columns):
        raise ValueError("`layer_links` must contain `from` and `to` columns.")

    if "weight" not in links.columns:
        links["weight"] = 1.0

    if ((links[["from", "to"]] < 1).any().any()) or ((links[["from", "to"]] > n_layers).any().any()):
        raise ValueError("`layer_links` indices must be between 1 and number of layers.")

    return links[["from", "to", "weight"]]


def _communities_to_membership(communities: Iterable[set[int]]) -> dict[int, int]:
    membership: dict[int, int] = {}
    for idx, nodes in enumerate(communities, start=1):
        for n in nodes:
            membership[n + 1] = idx
    return membership


def fit_layer_communities(
    graph_layers: list[nx.Graph | nx.DiGraph],
    algorithm: str = "louvain",
    resolution_parameter: float = 1.0,
    directed: bool = False,
) -> list[LayerCommunityFit]:
    algorithm = algorithm.lower()
    if algorithm not in {"louvain", "leiden"}:
        raise ValueError("`algorithm` must be one of {'louvain', 'leiden'}.")

    fits: list[LayerCommunityFit] = []

    for g in graph_layers:
        if algorithm == "louvain":
            try:
                import community as community_louvain
            except ImportError as exc:
                raise ImportError("Install optional dependency `python-louvain` for Louvain support.") from exc

            g_input = g.to_undirected() if directed else g
            partition = community_louvain.best_partition(g_input, weight="weight", resolution=resolution_parameter)
            communities = {}
            for node_zero, comm in partition.items():
                communities.setdefault(comm, []).append(node_zero + 1)

            mod = None
            if not directed:
                sets = [set(n - 1 for n in nodes) for nodes in communities.values()]
                mod = nx.algorithms.community.modularity(g_input, sets, weight="weight")

            fits.append(LayerCommunityFit(membership={k + 1: v + 1 for k, v in partition.items()}, modularity=mod, communities={k + 1: v for k, v in communities.items()}))
        else:
            try:
                import igraph as ig
                import leidenalg
            except ImportError as exc:
                raise ImportError("Install optional dependencies `python-igraph` and `leidenalg` for Leiden support.") from exc

            g_input = g
            edges = [(u, v) for u, v in g_input.edges()]
            weights = [g_input[u][v].get("weight", 1.0) for u, v in edges]
            ig_graph = ig.Graph(
                n=g_input.number_of_nodes(),
                edges=edges,
                directed=g_input.is_directed(),
            )
            if weights:
                ig_graph.es["weight"] = weights

            objective = leidenalg.CPMVertexPartition if directed else leidenalg.ModularityVertexPartition
            partition = leidenalg.find_partition(
                ig_graph,
                objective,
                weights=weights if weights else None,
                resolution_parameter=resolution_parameter,
            )
            membership = {idx + 1: comm + 1 for idx, comm in enumerate(partition.membership)}
            comms = {i + 1: [node + 1 for node in sorted(nodes)] for i, nodes in enumerate(partition)}

            mod = None if directed else ig_graph.modularity(partition.membership, weights=weights if weights else None)
            fits.append(LayerCommunityFit(membership=membership, modularity=mod, communities=comms))

    return fits


def weighted_jaccard(a: list[int], b: list[int]) -> float:
    inter = len(set(a).intersection(b))
    union = len(set(a).union(b))
    return 0.0 if union == 0 else inter / union


def weighted_overlap(a: list[int], b: list[int]) -> float:
    inter = len(set(a).intersection(b))
    min_size = min(len(a), len(b))
    return 0.0 if min_size == 0 else inter / min_size


def community_overlap_edges(
    fit: list[LayerCommunityFit],
    layer_links: pd.DataFrame,
    metric: str = "jaccard",
    min_similarity: float = 0.0,
) -> pd.DataFrame:
    if metric not in {"jaccard", "overlap"}:
        raise ValueError("`metric` must be one of {'jaccard', 'overlap'}.")

    sim_fun = weighted_jaccard if metric == "jaccard" else weighted_overlap
    rows: list[dict] = []

    for _, row in layer_links.iterrows():
        from_idx = int(row["from"])
        to_idx = int(row["to"])
        layer_weight = float(row["weight"])

        from_comms = fit[from_idx - 1].communities
        to_comms = fit[to_idx - 1].communities

        for from_c, from_nodes in from_comms.items():
            for to_c, to_nodes in to_comms.items():
                sim = sim_fun(from_nodes, to_nodes)
                weighted_sim = sim * layer_weight
                if weighted_sim >= min_similarity:
                    rows.append(
                        {
                            "from_layer": from_idx,
                            "to_layer": to_idx,
                            "from_community": from_c,
                            "to_community": to_c,
                            "similarity": sim,
                            "layer_weight": layer_weight,
                            "weighted_similarity": weighted_sim,
                        }
                    )

    return pd.DataFrame(rows)
